Mathlib/RingTheory/Ideal/KrullHeightTheorem.lean

```diff
@@ -0,0 +1,300 @@
+    rw [wellFoundedGT_dual_iff]; exact this) qs
+  refine ⟨n, (?_ : q.map f ^ n = 0)⟩
+  apply Submodule.eq_bot_of_le_smul_of_le_jacobson_bot (q.map f) _ (IsNoetherian.noetherian _)
+  rotate_left
+  · rw [IsLocalRing.jacobson_eq_maximalIdeal, Localization.AtPrime.map_eq_maximalIdeal]
+    exact bot_ne_top
+  rw [smul_eq_mul, ← pow_succ',
+    ← IsLocalization.map_comap q.primeCompl (Localization.AtPrime q) (q.map f ^ n),
+    ← IsLocalization.map_comap q.primeCompl (Localization.AtPrime q) (q.map f ^ (n + 1))]
+  apply Ideal.map_mono
+  apply Submodule.le_of_le_smul_of_le_jacobson_bot (IsNoetherian.noetherian _) (_ : I ≤ _)
+  swap
+  · rw [IsLocalRing.jacobson_eq_maximalIdeal]
+    exacts [hp.1.2, bot_ne_top]
+  · specialize hn _ n.le_succ
+    apply_fun Ideal.comap (Ideal.Quotient.mk (I := I)) at hn
+    simp only [qs, OrderHom.coe_mk, ← RingHom.ker_eq_comap_bot, Ideal.mk_ker,
+      Ideal.comap_map_of_surjective (Ideal.Quotient.mk (I := I)) Ideal.Quotient.mk_surjective] at hn
+    intro x hx
+    obtain ⟨y, hy, z, hz, rfl⟩ := Submodule.mem_sup.mp ((hn.le : _) (Ideal.mem_sup_left hx))
+    refine Submodule.add_mem_sup hy ?_
+    obtain ⟨z, rfl⟩ := (Submodule.IsPrincipal.mem_iff_eq_smul_generator I).mp hz
+    rw [smul_eq_mul, smul_eq_mul, mul_comm]
+    refine Ideal.mul_mem_mul ?_ (Submodule.IsPrincipal.generator_mem _)
+    rwa [Ideal.mem_comap, f.map_add, smul_eq_mul, f.map_mul, Ideal.add_mem_iff_right _
+      (Ideal.pow_le_pow_right n.le_succ hy), mul_comm, Ideal.unit_mul_mem_iff_mem] at hx
+    refine IsLocalization.map_units _ ⟨_,
+      show Submodule.IsPrincipal.generator I ∈ q.primeCompl from ?_⟩
+    show Submodule.IsPrincipal.generator I ∉ (↑q : Set R)
+    rw [← Set.singleton_subset_iff, ← Ideal.span_le, Ideal.span_singleton_generator]
+    exact fun e => h₂.not_le (hp.2 ⟨h₁, e⟩ h₂.le)
+
+/-- Krull's Hauptidealsatz: In a commutative Noetherian ring `R`,
+  any prime ideal that is minimal over a principal ideal has height at most 1. -/
+lemma Ideal.height_le_one_of_isPrincipal_of_mem_minimalPrimes [IsNoetherianRing R]
+    (I : Ideal R) (hI : I.IsPrincipal) (p : Ideal R) (hp : p ∈ I.minimalPrimes) : p.height ≤ 1 := by
+  haveI := hp.1.1
+  cases subsingleton_or_nontrivial R with
+  | inl h =>
+    exact (‹p.IsPrime›.ne_top (Subsingleton.elim _ _)).elim
+  | inr h =>
+    let f := algebraMap R (Localization.AtPrime p)
+    have := Ideal.height_le_one_of_isPrincipal_of_mem_minimalPrimes_of_isLocalRing
+      (I.map f) ⟨⟨f (Submodule.IsPrincipal.generator I), ?_⟩⟩ ?_
+    { rwa [← IsLocalization.height_comap p.primeCompl,
+        Localization.AtPrime.comap_maximalIdeal] at this }
+    { rw [← Set.image_singleton, ← Ideal.span, ← Ideal.map_span, Ideal.span_singleton_generator I] }
+    { rwa [IsLocalization.minimalPrimes_map p.primeCompl (Localization.AtPrime p) I,
+        Set.mem_preimage, Localization.AtPrime.comap_maximalIdeal] }
+
+set_option linter.style.multiGoal false in
```

Comment:
This shouldn't be here. Please refactor the proof instead of disabling this linter.